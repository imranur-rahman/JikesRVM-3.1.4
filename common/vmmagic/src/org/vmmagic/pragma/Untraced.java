/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;

/**
 * Use this annotation to mark fields that are read without the knowledge of the
 * memory management system. This means that barriers are not triggered and
 * the reference is also not traced by the garbage collector.<p>
 *
 * In order to guarantee that the constraints described above are honored by the compiler,
 * all accesses to untraced fields must be made when the field is resolved. If the field
 * were not resolved at the time when the code for the access was being compiled, the
 * Untraced annotation wouldn't be available (annotations are loaded at resolution
 * time).
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.FIELD)
public @interface Untraced { }
