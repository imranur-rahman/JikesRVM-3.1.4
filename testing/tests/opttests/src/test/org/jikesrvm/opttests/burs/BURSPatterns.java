/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.opttests.burs;

import org.vmmagic.pragma.NoInline;

/**
 * Tests to cover some BURS patterns.
 * <p>
 * The current tests revealed bugs in the 32-bit version of the IA32 LONG_MUL
 * BURS rules after the 32-bit / 64-bit split for the IA32 rules.
 * <p>
 * If you want to delete (tests from) this class, please consider if
 * the tests can be turned into a generally useful optimizing compiler
 * test.
 */
public class BURSPatterns {

  public static void main(String[] args) {
    System.out.println("signed_32bit_32bit_to_64bit_integer_multiplication");
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(0, 0));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(0, 1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(1, 1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(1, -1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(-1, 1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(-1, Integer.MIN_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(-1, Integer.MAX_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(Integer.MAX_VALUE, Integer.MAX_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(3, Integer.MAX_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication(3, Integer.MIN_VALUE));

    System.out.println("signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant");
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(0));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(-1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(-2));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(2));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(Integer.MAX_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(Integer.MIN_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(-5));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(7));

    System.out.println("signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant");
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(0));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(-1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(1));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(-2));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(2));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(Integer.MAX_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(Integer.MIN_VALUE));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(-5));
    System.out.println(signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(7));
  }

  @NoInline
  private static long signed_32bit_32bit_to_64bit_integer_multiplication(int a, int b) {
    return a * b;
  }

  @NoInline
  private static long signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant(int a) {
    return a * 0x00000000FFFFFFFFL;
  }

  @NoInline
  private static long signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant(int a) {
    return a * 0xFFFFFFFF0000FDFDL;
  }

}
