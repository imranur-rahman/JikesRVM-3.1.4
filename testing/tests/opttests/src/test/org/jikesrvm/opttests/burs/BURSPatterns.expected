signed_32bit_32bit_to_64bit_integer_multiplication
0
0
1
-1
-1
-2147483648
-2147483647
1
2147483645
-2147483648
signed_32bit_32bit_to_64bit_integer_multiplication_pos_long_constant
0
-4294967295
4294967295
-8589934590
8589934590
9223372030412324865
-9223372034707292160
-21474836475
30064771065
signed_32bit_32bit_to_64bit_integer_multiplication_neg_long_constant
0
4294902275
-4294902275
8589804550
-8589804550
-9223232401025596925
9223232405320499200
21474511375
-30064315925
